### DISCUSSION TITLE
AspNet: How to reuse NGINX request_id as TraceId

### DISCUSSION THREAD

**Discussion by geemark** [open-telemetry/opentelemetry-dotnet#3943] [Q&A] [answered]
I have a NGINX load balancer in front of my applications, and I want to use the request_id generated from NGINX as the Trace ID when stating both AspNet and AspNetCore instrumentation. The ID is forwarded in a custom header from nginx and is already a 32 char hex string, so it should be possible.

I was thinking of using AspNetCoreInstrumentationOptions.EnrichWithHttpRequest for .NET and AspNetInstrumentationOptions.Enrich for .NET Framework. However, the TraceId is read only, presumably because it must be immutable after the activity is created.

Is there a way to set a custom TraceId when the activity gets first created for a web request?

**Comment by vishweshbankwar** [comment, marked as answer]
@geemark - Have you considered using custom propagator for this?

**Reply by geemark** [reply]
I am very new to OpenTelemetry, so I am not aware of that concept.
Can you give me a few quick pointers to get started? Link to docs, or maybe just the name of the interface I need to implement, and how to add it to the Sdk builder?

**Reply by vishweshbankwar** [reply]
You can use Sdk.SetDefaultTextMapPropagator to set your custom propagator. Take a look at one of the propagators shipped from this repo on how to implement. https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Extensions.Propagators

**Reply by geemark** [reply]
Thank you for your suggestion. It seems I have a working solution now :)
